### DISCUSSION TITLE
Requirements Engineering (RE)

### DISCUSSION THREAD

**Discussion by Lica200** [SE-TINF22B2/G4-Get2Gether#23] [Documentation (deprecated)]
# Fachliche Ebene
(hier steht die reine Fachlichkeit - Wie entsteht ein _Event_ (vgl. #40):
Unternehmungen in der Gruppe werden oft in Chatgruppen organisiert. Ein Organisator gibt dann die Grundlegende Idee und den Ort des Events bekannt und es folgen Terminabstimmung (wenn dieser nicht fest ist) und andere Aktionen. Unter die anderen Aktionen fallen:
- Kommentare wie das Projekt wahrgenommen wird und Nachfragen
- Einkaufslisten (“was muss eingekauft/mitgebracht werden, um das Projekt zu realisieren?”) mit der Möglichkeit anzugeben, welche Person für was zuständig ist.
- Fahrgemeinschaften (“wer kann Andere im Auto mitnehmen?”) - wenn nötig

Nach dem Event folgen oft noch die folgenden Aktionen:
- Ausgaben anpassen (Personen können hier angeben was sie ausgegeben haben, und die Kosten können ausgeglichen werden)
- Fotos werden ausgetauscht
# Anforderungen aus Anwendersicht
(hier wird angegeben welcher Teil der fachlichen Ebene durch die Webapp umgesetzt werden soll)
- Es soll eine Website umgesetzt werden, in der Events erstellt und gespeichert werden können.
- Die Website umfasst ein Dashboard, in dem Events angezeigt werden, ein Formular zum Erstellen von Events und Seiten für Registrierung und Login.
### Projektmanagement
- Anforderungen müssen formuliert werden.
- Ein Prototyp für das Design der Website muss erstellt werden.
- Abnahmekriterien müssen bestimmt werden, anhand derer der der Erfolg des Projekts überprüft werden kann.
- Die benutzten Technologien müssen begründet ausgewählt werden.
### Datenbank
- Die Daten der Events müssen in einer Datenbank gespeichert werden. Die Attribute und Datentypen müssen dafür definiert werden.
- Die Benutzer sollen in einer Datenbank gespeichert werden. Welche Daten hierfür nötig sind, soll dafür in einem ER-Modell festgehalten werden.
- Die von den Widgets benötigten Attribute und Datentypen sollen in einem Grunddatenmodell festgehalten werden.
### Event
- ![Use Case Diagram1](https://github.com/SE-TINF22B2/G4-Get2Gether/assets/134512122/34772bb6-4754-4a59-b787-6b57bacd4ece)

  - Bevor ein Event oder Link erstellt werden kann, wird überprüft, ob der Benutzer die Befugnisse dazu hat.
  - Beispiel Event erstellen: Der ersteller gibt in der Webapp die Daten für das Event ein (z.B. Name und Beschreibung), daraufhin wird dieses an das Backend weitergegeben. Dort wird das Event überprüft und wenn es keine Fehler gibt (z.B. leeres Feld) angelegt.
  - Beispiel Link erstellen: Um das Event mit Freunden teilen zu können, muss der Ersteller einen Link erstellen. Dafür wird eine Anfrage an das Backend geschickt, die den Link zurück gibt.
- Als Teilnehmer möchte ich durch den Link des Erstellers in die Webapp geleitet werden und dort zunächst Eventnamen und Eventbeschreibung sehen.
- Events erlauben ein Einbinden von Widgets.
### Login
- Als Benutzer will ich durch Benutzername und Passwort ein eigenes Konto haben.
- Nur nach erfolgreichem Einloggen kann auf <<...>> zugegriffen werden.
- Als Benutzer will ich auf die Events zugreifen können, die ich bereits erstellt habe.
### Widget-Anforderungen
Hier werden die _Widgets_ (vgl. #40) aufgelistet, die während der Entwicklungszeit der Website hinzugefügt werden. Widgets sind dabei Funktionalitäten, die zusätzlich zu der Grundfunktionalität (Events erstellen und Freunde einladen) bestehen. Die Reihenfolge, in der die Widgets aufgelistet sind, spiegelt auch die Reihenfolge der Umsetzung wieder. Nicht alle Widgets werden realisiert, die ist eine Auflistung, was im Rahmen des Projekts möglich ist.

Einkaufsliste:
Es soll eine Einkaufsliste hinzugefügt werden, bei der sich die Teilnehmer des Events eintragen können. 

Karte (Google Maps):
Es wird ein Kartenausschnitt der Adresse des Eventortes dargestellt.

Abstimmung:
Hier können Teilnehmer Abstimmungen erstellen. Dies kann zum Beispiel zur Terminfindung verwendet werden. Jede Abstimmung enthält einen Titel, eine Frage und mehrere Antwortmöglichkeiten. Jeder Eventteilnehmer kann die Ergebnisse in (nahezu) Echtzeit einsehen und für jede Antwortmöglichkeit kann mehrfach ausgewählt werden (aber von jeder Person nur einmal).

Fahrgemeinschaft.
Teilnehmer sollen sich als Fahrer und Mitfahrer eintragen können und somit Fahrgemeinschaften bilden.
Ein Fahrer kann ein Auto erstellen und dabei die Anzahl der Sitzplätze angeben. Entsprechend viele Mitfahrer können sich in dieses Auto eintragen.

Ausgabenverteilung:
Teilnehmer können gezahlte Ausgaben hinzufügen, welche dann auf alle Teilnehmer verteilt werden können (ähnlich Splitwise).

Galerie (Google Drive):
In diesem Widget können alle Teilnehmer Bilder zum erlebten Event ohne Qualitätsverlust austauschen.

Wetter (OpenWeatherMap):
In diesem Widget werden die Wetterdaten für den Zeitraum und Ort des Events angezeigt. Die Wetterdaten werden von der OpenWeatherMap-API abgerufen.

Kommentarfunktion:
Die Teilnehmer können Kommentare in einem Chat austauschen. Die Kommentare sind einfache Textnachrichten.

Optionale Features
- Gruppen erstellen und einladen
- Dark Mode
- Sprache Englisch-Deutsch

**Comment by roland-KA** [comment]
Heute in der Vorlesung war die Zeit etwas zu knapp für ausführlicheres Feedback. Deshalb hier noch ein paar Zeilen dazu:
- Da hier eigentlich sämtliche Aspekte der rein fachlichen Ebene auch in Software abgebildet werden sollen, würde ich einfach die erste Überschrift in "Einführung" umbenennen und den Absatz schließen mit einem Satz wie "All diese Aktivitäten sollen durch eine Web-basierte Anwendung unterstützt werden, die im folgenden näher spezifiziert wird".
- Im nächsten Abschnitt wird die Funktionalität aus Anwendersicht beschrieben. Da wäre es gut, wenn Sie auf Begriffe wie "Dashboard" und "Widgets" verzichten könnten 🤓.
  - Ein Anwender interessiert sie nach dem einleitenden Abschnitt vermutlich zunächst einmal dafür, wie er sich und seine Freunde als Gruppe anlegen kann und wie er im nächsten Schritt Events für diese Gruppe erstellen kann.
  - Danach macht es Sinn, die Möglichkeiten zu erläutern, die die Anwendung für die Organisation und Durchführung des Events anbietet.
- Die unter "Projektmanagement" aufgeführten Punkte sind einfach Tasks, die Sie in Ihrer Projektplanung unterbringen. Die gehören nicht in das Anforderungsdokument.
- Die Aspekte unter "Datenbank" sind schon relativ technisch und umsetzungsnah. In den Anforderungen aus Anwendersicht soll natürlich hervorgehen, dass bestimmte Daten gespeichert werden. Das ist dort jedoch eher ein impliziter Aspekt. Wenn Sie (in weiteren Abschnitten) die Funktionalitäten aus Anwendersicht weiter konkretisieren (z.B. mit Mockups) geht das Dokument immer mehr in den Bereich "Pflichtenheft" über. In dem Kontext ist auch ein fachliches Datenmodell angebracht.
